### PR TITLE
Pull `HSI_Event` class into its own submodule

### PR DESCRIPTION
**Pull request by willGraham01**
This pull request moves the `HSI_Event` class (and related supporting classes) out of the `healthsystem` module and into its own submodule (`hsi_event`). This allows us to:

- Add typehints to the `HSI_Event` class.
- Perform some simplifications to the internal class logic (notably not storing objects that are not referenced, or can be obtained through properties rather than storing copies of the original object).
- These changes are largely intended to help make implementing changes for #1237 simpler and slightly less noisy.
 
There are no functionality changes that are introduced by this PR.

## Key changes

- The `HSI_Event` class and related container classes have been moved into the `src/tlo/methods/hsi_event.py` file, to make them accessible without needing to import the entire `healthcaresystem.py` module.
  - This also lets us avoid circular imports that previously prevented type-hinting.
- Internal logic of the `HSI_Event` class has been simplified slightly in places:
  - Class attributes are given typehints.
  - Avoids storing the parent `Module`, `Simulation`, and `Population`.
    - We now just store the `Module`.
    - The simulation can be accessed through a `self.sim` property to avoid storing it as a class attribute.
    - The healthcare system can also be accessed through a shortcut `self.healthsystem` property.
    - The `target_is_alive` property is used in place of tracking the population dataframe as a class attribute. We were only looking up one value from the dataframe, and otherwise not using it at all.
  - Avoids dynamically creating the `return_item_codes_in_dict` function every time the `get_consumables` method is run.
    - `return_item_codes_in_dict` is now a stand-alone method.
  - `adjust_facility_level_to_merge_1b_and_2` has been made into a class method since it is not used outside of the `HSI_Event` class.
  - Typehints have been added to member functions and attributes where possible.

### CODE REVIEW

**Review by matt-graham — APPROVED** [review]
Thanks @willGraham01 these all looks like sensible changes to make. I have made a few small suggestions though none are vital. 

Just so its recorded somewhere, Will and I discussed in person the changes to the imports of `HSI_Event` in other modules, which are not strictly necessary as `HSI_Event` is imported from the new `tlo.methods.hsi_event` module in `tlo.methods.healthsystem` and so could technically still be imported from there, which would keep the changes in this PR confined to the `healthsystem` and new `hsi_event` modules. Will pointed out changing the imports helps avoid circular imports between disease modules and healthsystem and should simplify importing `HSI_Event` to use just for type hinting purposes without needing any `if TYPE_CHECKING` blocks, which I felt was a good enough justification to outweigh the increased probability of needing to deal with some merge conflicts in disease modules on other branches when this is merged in (though as this only touches imports shouldn't be too much of a problem).

File: src/tlo/methods/hsi_event.py [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+                for code, quantity in item_codes.items()
+            ]
+        ):
+            raise ValueError(
+                "item_codes must be integers and quantities must be integers or floats."
+            )
+        return {int(i): float(q) for i, q in item_codes.items()}
+
+    else:
+        raise ValueError("The item_codes are given in an unrecognised format")
```
[the remark this answers is not in the document]
Comment (willGraham01) [reply src/tlo/methods/hsi_event.py]: Have converted it to a `@staticmethod` and shifted it into `HSI_Event`. It was previously in here, but was being re-defined within the `get_consumables` method each time it ran

**Comment by willGraham01** [comment]
Have adapted the type-hinting to use `from future` and moved the stray method into the `HSI_Event` class to hide it. Re-requesting review in case there's anything else, otherwise will merge when (/if!) tests pass

**Review by matt-graham — APPROVED** [review]
Thanks for doing the updates

**Pull request merged**